From 32eb9bcb93e8308cc4ac60652accb500279716f6 Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Tue, 24 Jun 2025 13:35:15 +0800
Subject: [PATCH 27/63] loonggpu: Use multihop

Remove the code to move resources directly between
SYSTEM and VRAM in favour of using the core ttm mulithop code.

Following the reference implementation change and various fixes after
it.

Link: https://git.kernel.org/torvalds/c/f5a89a5cae81
Link: https://git.kernel.org/torvalds/c/aefec40938e4
Link: https://git.kernel.org/torvalds/c/c92db8d64f9e
Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 loonggpu/loonggpu_ttm.c | 178 ++++++----------------------------------
 1 file changed, 25 insertions(+), 153 deletions(-)

diff --git a/loonggpu/loonggpu_ttm.c b/loonggpu/loonggpu_ttm.c
index 6bc7d99..cc1cac1 100644
--- a/loonggpu/loonggpu_ttm.c
+++ b/loonggpu/loonggpu_ttm.c
@@ -441,7 +441,7 @@ error:
 /**
  * loonggpu_move_blit - Copy an entire buffer to another buffer
  *
- * This is a helper called by loonggpu_bo_move() and loonggpu_move_vram_ram() to
+ * This is a helper called by loonggpu_bo_move() to
  * help move buffers to and from VRAM.
  */
 static int loonggpu_move_blit(struct ttm_buffer_object *bo,
@@ -479,137 +479,6 @@ error:
 	return r;
 }
 
-/**
- * loonggpu_move_vram_ram - Copy VRAM buffer to RAM buffer
- *
- * Called by loonggpu_bo_move().
- */
-static int loonggpu_move_vram_ram(struct ttm_buffer_object *bo, bool evict,
-				struct ttm_operation_ctx *ctx,
-				lg_ttm_mem_t *new_mem)
-{
-	struct loonggpu_device *adev;
-	lg_ttm_mem_t *old_mem = lg_tbo_to_mem(bo);
-	lg_ttm_mem_t tmp_mem, *p;
-	struct ttm_place placements;
-	struct ttm_placement placement;
-	int r;
-
-	adev = loonggpu_ttm_adev(bo->bdev);
-
-	/* create space/pages for new_mem in GTT space */
-	tmp_mem = *new_mem;
-	p = &tmp_mem;
-	lg_clear_drm_node(lg_res_to_drm_node(&tmp_mem));
-	placement.num_placement = 1;
-	placement.placement = &placements;
-	lg_ttm_placement_set_busy_place(&placement, &placements, 1);
-	placements.fpfn = 0;
-	placements.lpfn = 0;
-#if defined(TTM_PL_FLAG_TT)
-#if defined(TTM_PL_MASK_CACHING)
-	placements.flags = TTM_PL_MASK_CACHING | TTM_PL_FLAG_TT;
-#else
-	placements.flags = TTM_PL_FLAG_TT;
-#endif
-#else
-#if defined(TTM_PL_MASK_CACHING)
-	placements.flags = TTM_PL_MASK_CACHING;
-#endif
-	placements.mem_type = TTM_PL_TT;
-#endif
-	r = lg_ttm_bo_mem_space(bo, &placement, p, ctx);
-	if (unlikely(r)) {
-		return r;
-	}
-
-#if defined(TTM_PL_FLAG_WC)
-	/* set caching flags */
-	r = ttm_tt_set_placement_caching(bo->ttm, tmp_mem.placement);
-	if (unlikely(r)) {
-		goto out_cleanup;
-	}
-#endif
-	/* Bind the memory to the GTT space */
-	r = lg_ttm_tt_bind(bo, &tmp_mem, ctx);
-	if (unlikely(r)) {
-		goto out_cleanup;
-	}
-
-	/* blit VRAM to GTT */
-	r = loonggpu_move_blit(bo, evict, ctx->no_wait_gpu, &tmp_mem, old_mem);
-	if (unlikely(r)) {
-		goto out_cleanup;
-	}
-
-	/* move BO (in tmp_mem) to new_mem */
-	r = lg_ttm_bo_move_ttm(bo, ctx, new_mem);
-out_cleanup:
-	lg_ttm_bo_mem_put(bo, &tmp_mem);
-	return r;
-}
-
-/**
- * loonggpu_move_ram_vram - Copy buffer from RAM to VRAM
- *
- * Called by loonggpu_bo_move().
- */
-static int loonggpu_move_ram_vram(struct ttm_buffer_object *bo, bool evict,
-				struct ttm_operation_ctx *ctx,
-				lg_ttm_mem_t *new_mem)
-{
-	struct loonggpu_device *adev;
-	lg_ttm_mem_t *old_mem = lg_tbo_to_mem(bo);
-	lg_ttm_mem_t tmp_mem, *p;
-	struct ttm_placement placement;
-	struct ttm_place placements;
-	int r;
-
-	adev = loonggpu_ttm_adev(bo->bdev);
-
-	/* make space in GTT for old_mem buffer */
-	tmp_mem = *new_mem;
-	p = &tmp_mem;
-	lg_clear_drm_node(lg_res_to_drm_node(&tmp_mem));
-	placement.num_placement = 1;
-	placement.placement = &placements;
-	lg_ttm_placement_set_busy_place(&placement, &placements, 1);
-	placements.fpfn = 0;
-	placements.lpfn = 0;
-#if defined(TTM_PL_FLAG_TT)
-#if defined(TTM_PL_MASK_CACHING)
-	placements.flags = TTM_PL_MASK_CACHING | TTM_PL_FLAG_TT;
-#else
-	placements.flags = TTM_PL_FLAG_TT;
-#endif
-#else
-	placements.mem_type = TTM_PL_TT;
-#if defined(TTM_PL_MASK_CACHING)
-	placements.flags = TTM_PL_MASK_CACHING;
-#endif
-#endif
-
-	r = lg_ttm_bo_mem_space(bo, &placement, p, ctx);
-	if (unlikely(r)) {
-		return r;
-	}
-
-	/* move/bind old memory to GTT space */
-	r = lg_ttm_bo_move_ttm(bo, ctx, &tmp_mem);
-	if (unlikely(r)) {
-		goto out_cleanup;
-	}
-
-	/* copy to VRAM */
-	r = loonggpu_move_blit(bo, evict, ctx->no_wait_gpu, new_mem, old_mem);
-	if (unlikely(r)) {
-		goto out_cleanup;
-	}
-out_cleanup:
-	lg_ttm_bo_mem_put(bo, &tmp_mem);
-	return r;
-}
-
 static lg_ttm_backend_unbind_ret loonggpu_ttm_backend_unbind(lg_ttm_backend_func_arg);
 
 /**
@@ -670,37 +539,40 @@ static int loonggpu_bo_move(struct ttm_buffer_object *bo, bool evict,
 		return 0;
 	}
 
-	if (!adev->mman.buffer_funcs_enabled)
-		goto memcpy;
+	if (bo->type == ttm_bo_type_device &&
+	    new_mem->mem_type == TTM_PL_VRAM &&
+	    old_mem->mem_type != TTM_PL_VRAM) {
+		/* loonggpu_bo_fault_reserve_notify will re-set this if the CPU
+		 * accesses the BO after it's moved.
+		 */
+		abo->flags &= ~LOONGGPU_GEM_CREATE_CPU_ACCESS_REQUIRED;
+	}
 
-	if (old_mem->mem_type == TTM_PL_VRAM &&
-	    new_mem->mem_type == TTM_PL_SYSTEM) {
-		r = loonggpu_move_vram_ram(bo, evict, ctx, new_mem);
-	} else if (old_mem->mem_type == TTM_PL_SYSTEM &&
-		   new_mem->mem_type == TTM_PL_VRAM) {
-		r = loonggpu_move_ram_vram(bo, evict, ctx, new_mem);
-	} else {
-		r = loonggpu_move_blit(bo, evict, ctx->no_wait_gpu,
-				     new_mem, old_mem);
+	if (adev->mman.buffer_funcs_enabled &&
+	    ((old_mem->mem_type == TTM_PL_SYSTEM &&
+	      new_mem->mem_type == TTM_PL_VRAM) ||
+	     (old_mem->mem_type == TTM_PL_VRAM &&
+	      new_mem->mem_type == TTM_PL_SYSTEM))) {
+		hop->fpfn = 0;
+		hop->lpfn = 0;
+		hop->mem_type = TTM_PL_TT;
+		hop->flags = 0;
+		return -EMULTIHOP;
 	}
 
+	if (adev->mman.buffer_funcs_enabled)
+		r = loonggpu_move_blit(bo, evict, ctx->no_wait_gpu, new_mem,
+				    old_mem);
+	else
+		r = -ENODEV;
+
 	if (r) {
-memcpy:
 		r = ttm_bo_move_memcpy(bo, ctx, new_mem);
 		if (r) {
 			return r;
 		}
 	}
 
-	if (bo->type == ttm_bo_type_device &&
-	    new_mem->mem_type == TTM_PL_VRAM &&
-	    old_mem->mem_type != TTM_PL_VRAM) {
-		/* loonggpu_bo_fault_reserve_notify will re-set this if the CPU
-		 * accesses the BO after it's moved.
-		 */
-		abo->flags &= ~LOONGGPU_GEM_CREATE_CPU_ACCESS_REQUIRED;
-	}
-
 out:
 	/* update statistics */
 	atomic64_add(lg_gbo_to_gem_obj(abo).size, &adev->num_bytes_moved);
-- 
2.52.0

